modules/social_features/social_group/modules/social_group_secret/social_group_secret.module

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/**
+ * @file
+ * The Social group secret module.
+ */
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\views\Plugin\views\query\QueryPluginBase;
+use Drupal\views\Plugin\views\row\EntityRow;
+use Drupal\views\ViewExecutable;
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ *
+ * Prohibit changing group type.
+ */
+function social_group_secret_form_group_secret_group_edit_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  foreach (array_keys($form['group_type']['#options']) as $type) {
+    $form['group_type'][$type]['#disabled'] = TRUE;
+  }
+
+  $id = array_search('_social_group_type_edit_submit', $form['actions']['submit']['#submit']);
+  unset($form['actions']['submit']['#submit'][$id]);
+}
+
+/**
+ * Implements hook_views_query_alter().
+ *
+ * Hide secret groups everywhere when the current user is not a member.
+ */
+function social_group_secret_views_query_alter(ViewExecutable $view, QueryPluginBase $query) {
+  if (empty($view->rowPlugin) || !($view->rowPlugin instanceof EntityRow) || $view->rowPlugin->getEntityTypeId() !== 'group') {
+    return;
+  }
+
+  $account = \Drupal::currentUser();
+
+  if ($account->hasPermission('create secret_group group')) {
+    return;
+  }
+
+  /** @var \Drupal\group\GroupMembershipLoaderInterface $service */
+  $service = \Drupal::service('group.membership_loader');
+
+  $ids = [];
```

Comment:
It would be good to add some comments here on why we're creating an `$ids` array and why we loop through the queries and add extra clauses.